prov/util/src/util_coll.c

```diff
@@ -1003,17 +1021,17 @@ int ofi_join_collective(struct fid_ep *ep, fi_addr_t coll_addr,
 		coll_mc = (struct util_coll_mc*) ((uintptr_t) coll_addr);
 	}
 
-	ret = util_coll_mc_alloc(&new_coll_mc);
-	if (ret)
-		return ret;
+	new_coll_mc = av_set->coll_mc;
+	if (!new_coll_mc) {
+		ret = util_coll_mc_alloc(&new_coll_mc);
+		if (ret)
+			goto err1;
+	}
```

Comment:
We always need to allocate a new util_coll_mc here.  This structure is returned to the caller through the mc parameter and must be closed by the app via fi_close().  Along this line, when we allocate the util_coll_mc, we should increment a reference count on the av_set.  Closing the mc should decrement the reference count.  I see that av_set has a ref count available, but I don't see that it's getting incremented by this function.  The decrement should be in util_coll_close().  Since the util_coll_mc also references the endpoint, references should be acquired and released on the endpoint as well.

You mentioned this, but the av_set->av_set_fid.fid.ops are missing.  That's where an av_set_close() function would be set.  Close should check that the ref count is 0, and fail otherwise, before freeing the av_set.  To set this, look at util_coll_fi_ops as an example.  You just need to define a util_av_set_close() function that checks that the ref count is 0, and frees the av_set.  Then setup a static struct (similar to util_coll_fi_ops) that sets the function pointers.  Finally, assign av_set->av_set_fid.fid_ops = the_name_of_the_static_struct inside the av_set creation call below.

Reply:
can you be more specific on how to change the ref count of the endpoint?

Reply:
So, yeah, we can't do that, at least not easily.  Ignore that request.